pkg/ipcache/kvstore.go

```diff
@@ -292,15 +293,41 @@ func findShadowedCIDR(pair *identity.IPIdentityPair) (identity.NumericIdentity,
 	return IPIdentityCache.LookupByIP(cidrStr)
 }
 
-func ipIdentityWatcher(listeners []IPIdentityMappingListener) {
-	log.Info("Starting IP identity watcher")
+// IPIdentityWatcher is a watcher that will notify when IP<->identity mappings
+// change in the kvstore
+type IPIdentityWatcher struct {
+	backend kvstore.BackendOperations
+	stop    chan struct{}
+}
 
-	for {
+// NewIPIdentityWatcher creates a new IPIdentityWatcher using the specified
+// kvstore backend
+func NewIPIdentityWatcher(backend kvstore.BackendOperations) *IPIdentityWatcher {
+	watcher := &IPIdentityWatcher{
+		backend: backend,
+		stop:    make(chan struct{}),
+	}
 
-		watcher := kvstore.ListAndWatch("endpointIPWatcher", IPIdentitiesPath, 512)
+	return watcher
+}
+
+// Run starts the watcher and blocks waiting for events. When events are
+// received from the kvstore, All IPIdentityMappingListener are notified. The
+// function returns when IPIdentityWatcher.Close() is called. The watcher will
+// automatically restart as required.
+func (iw *IPIdentityWatcher) Run(listeners []IPIdentityMappingListener) {
```

Comment:
This might be better named `Watch` since it blocks, but I'm not sure why I thought `Run` would return so this might be specific to me.